datacube.py

```diff
@@ -0,0 +1,382 @@
+    """
+    Search for Digital Elevation Model (DEM) items within a given bounding box (BBOX), STAC catalog, week, and Sentinel-2 items.
+
+    Parameters:
+    - BBOX (tuple): Bounding box coordinates in the format (minx, miny, maxx, maxy).
+    - catalog (pystac.Catalog): STAC catalog containing DEM items.
+    - week (str): The week in the format 'start_date/end_date'.
+    - s2_items (list): List of Sentinel-2 items.
+
+    Returns:
+    - tuple: A tuple containing DEM items and DEM GeoDataFrame.
+    """
+    search = catalog.search(
+        collections=["cop-dem-glo-30"], bbox=BBOX
+    )
+    dem_items = search.item_collection()
+    print(f"Found {len(dem_items)} items")
+
+    dem_gdf = gpd.GeoDataFrame.from_features(dem_items.to_dict())
+
+    dem_gdf.set_crs(epsg=4326, inplace=True)
+    dem_gdf = dem_gdf.to_crs(epsg=s2_items[0].properties["proj:epsg"])
+    return dem_items, dem_gdf
+
+
+def make_dataarrays(s2_items, s1_items, dem_items, BBOX, resolution, epsg):
+    """
+    Create xarray DataArrays for Sentinel-2, Sentinel-1, and DEM data.
+
+    Parameters:
+    - s2_items (list): List of Sentinel-2 items.
+    - s1_items (list): List of Sentinel-1 items.
+    - dem_items (list): List of DEM items.
+    - BBOX (tuple): Bounding box coordinates in the format (minx, miny, maxx, maxy).
+    - resolution (int): Spatial resolution.
+    - epsg (int): EPSG code for the coordinate reference system.
+
+    Returns:
+    - tuple: A tuple containing xarray DataArrays for Sentinel-2, Sentinel-1, and DEM.
+    """
+    da_sen2: xr.DataArray = stackstac.stack(
+        items=s2_items[0],
+        epsg=26910,  # UTM Zone 10N
+        assets=S2_BANDS,
+        bounds_latlon=BBOX,  # W, S, E, N
+        resolution=10,  # Spatial resolution of 10 metres
+        xy_coords="center",  # pixel centroid coords instead of topleft corner
+        dtype=np.float32,
+        fill_value=np.nan,
+    )
+
+    da_sen1: xr.DataArray = stackstac.stack(
+        items=s1_items[1:],  # To only accept the same orbit state and date. Need better way to do this.
+        assets=["vh", "vv"],  # SAR polarizations
+        epsg=26910,  # UTM Zone 10N
+        bounds_latlon=BBOX,  # W, S, E, N
+        xy_coords="center",  # pixel centroid coords instead of topleft corner
+        dtype=np.float32,
+        fill_value=np.nan,
+    )
+
+    # To fix TypeError: Invalid value for attr 'spec'
+    da_sen1.attrs["spec"] = str(da_sen1.spec)
+
+    # To fix ValueError: unable to infer dtype on variable None
+    for key, val in da_sen1.coords.items():
+        if val.dtype == "object":
+            print("Deleting", key)
+            da_sen1 = da_sen1.drop_vars(names=key)
+
+    # Create xarray.Dataset datacube with VH and VV channels from SAR
+    da_vh: xr.DataArray = da_sen1.sel(band="vh", drop=True).rename("vh")
+    da_vv: xr.DataArray = da_sen1.sel(band="vv", drop=True).rename("vv")
+    ds_sen1: xr.Dataset = xr.merge(objects=[da_vh, da_vv], join="override")
```

Comment:
`da_sen1 = stackstac.mosaic(da_sen1, dim="time")` will create a composite of S1 tiles overlapping the S2 bbox. We can use that directly in the merge, and don't have to worry about pixel picking stratergy.

Reply:
Word. I like this. Thanks.

Reply:
So `stackstac.mosaic` returns an `xarray.DataArray`, whereas `xarray.merge` returns an `xarray.Dataset`. The main difference is that `xarray.Dataset` allows us to clearly show named data variables (e.g. VV, VH, B1, B2, etc), whereas `xarray.DataArray` stores these in the 'band' dimension. Do we have a preference for either format?

Reply:
Having named data vars helps - `stackstac.mosaic` takes care of merging the pixels along the given dim (time in our case) which we don't have to handle at our end.
What if we create the data arrays & then `concat` them together as a dataset with vars attached?

Reply:
We still retain the named data vars I think. Merged data array using mosaic method for Sentinel 1:

```
Merged datarray:  <xarray.Dataset>
Dimensions:                                     (time: 1, band: 14, x: 11365,
                                                 y: 11363)
Coordinates: (12/67)
  * time                                        (time) datetime64[ns] 2018-06...
  * band                                        (band) <U4 'B02' 'B03' ... 'vv'
  * x                                           (x) float64 2.973e+05 ... 4.1...
  * y                                           (y) float64 4.202e+06 ... 4.0...
    id                                          (time) <U54 'S2B_MSIL2A_20180...
    s2:generation_time                          <U24 '2020-10-12T02:49:45.926Z'
    ...                                          ...
    sar:instrument_mode                         <U2 'IW'
    sar:product_type                            <U3 'GRD'
    s1:product_timeliness                       <U8 'Fast-24h'
    s1:resolution                               <U4 'high'
    description                                 (band) object nan ... 'Terrai...
    proj:shape                                  object {3600}
Data variables:
    stackstac-95bc7f7975c3ca621cbffb094f33316e  (time, band, y, x) float32 dask.array<chunksize=(1, 1, 1024, 1024), meta=np.ndarray>
    stackstac-80df347b660c0b771e7ba977f16777bc  (band, y, x) float32 dask.array<chunksize=(13, 1024, 1024), meta=np.ndarray>
    stackstac-602439b1960e897b8044c8056eb3ce96  (band, y, x) float32 dask.array<chunksize=(14, 1024, 1024), meta=np.ndarray>
Attributes:
    spec:        RasterSpec(epsg=32611, bounds=(297340, 4088320, 410990, 4201...
    crs:         epsg:32611
    transform:   | 10.00, 0.00, 297340.00|\n| 0.00,-10.00, 4201950.00|\n| 0.0...
    resolution:  10
```